lucille-core/src/main/java/com/kmwllc/lucille/stage/OpenAIEmbedding.java

```diff
@@ -0,0 +1,194 @@
+import com.kmwllc.lucille.core.Document;
+import com.kmwllc.lucille.core.OpenAIModels;
+import com.kmwllc.lucille.core.Stage;
+import com.kmwllc.lucille.core.StageException;
+import com.knuddels.jtokkit.Encodings;
+import com.knuddels.jtokkit.api.Encoding;
+import com.knuddels.jtokkit.api.EncodingResult;
+import com.knuddels.jtokkit.api.ModelType;
+import com.typesafe.config.Config;
+import dev.langchain4j.data.segment.TextSegment;
+import java.util.ArrayList;
+import java.util.Iterator;
+import dev.langchain4j.data.embedding.Embedding;
+import dev.langchain4j.model.embedding.EmbeddingModel;
+import dev.langchain4j.model.openai.OpenAiEmbeddingModel;
+import dev.langchain4j.model.output.Response;
+import java.util.List;
+import java.util.Random;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This stage uses openAI embedding services to embed a text field in a Lucille document or its children document(s). Will truncate
+ * to token limit before sending request. Retrieves API key from environment variables.
+ *
+ * Config Parameters:
+ * - text_field (String) : field of which the embedding Stage will retrieve content from
+ * - embed_document (Boolean) : Embeds the document's text_field if set to true.
+ * - embed_children (Boolean): Embeds the document's children text_field if set to true.
+ * - api_key (String) : API key used for OpenAI requests
+ * - embedding_field (String, Optional) : name of the field that will hold the embeddings
+ * - model_name (String, Optional) : the name of the OpenAI embedding model to use, set default to text-embedding-3-small
+ *    1. text-embedding-3-small
+ *    2. text-embedding-3-large
+ *    3. text-embedding-ada-002
+ *    more details: <a href="https://platform.openai.com/docs/guides/embeddings/embedding-models">...</a>
+ * - dimensions (Integer, Optional) : number of dimensions the resulting embedding should have. Only supported in text-embedding-3
+ * and later models. Default set to null, which will call the model's default dimensions.
+ */
+
+public class OpenAIEmbedding extends Stage {
+  // this is the token limit for all embedding models from openai
+  private static final int DEFAULT_OPENAI_TOKEN_LIMIT = 8191;
+  private final String API_KEY;
+  private EmbeddingModel model;
+  private final String textField;
+  private final boolean embedDocument;
+  private final boolean embedChildren;
+  private final String embeddingField;
+  private final OpenAIModels modelName;
+  private final Integer dimensions;
+  private Encoding enc;
+
+  private static final Logger log = LoggerFactory.getLogger(OpenAIEmbedding.class);
+
+  public OpenAIEmbedding(Config config) throws StageException {
+    super(config, new StageSpec()
+        .withRequiredProperties("text_field", "embed_document", "embed_children", "api_key")
+        .withOptionalProperties("embedding_field", "model_name", "dimensions"));
+    textField = config.getString("text_field");
```

Comment:
convention is to use "this." prefix when setting instance variables inside a constructor

Reply:
resolved in commit 0c8bc21